confluent_kafka/avro/CachedSchemaRegistryClient.py

```diff
@@ -0,0 +1,288 @@
+            response = requests.get(url, headers=_headers)
+        elif method == 'POST':
+            response = requests.post(url, body, headers=_headers)
+        elif method == 'PUT':
+            response = requests.put(url, body, headers=_headers)
+        elif method == 'DELETE':
+            response = requests.delete(url, headers=_headers)
+        result = json.loads(response.text)
+        return (result, response.status_code)
+
+    def _add_to_cache(self, cache, subject, schema, value):
+        if subject not in cache:
+            cache[subject] = {}
+        sub_cache = cache[subject]
+        sub_cache[schema] = value
+
+    def _cache_schema(self, schema, schema_id, subject=None, version=None):
+        # don't overwrite anything
+        if schema_id in self.id_to_schema:
+            schema = self.id_to_schema[schema_id]
+        else:
+            self.id_to_schema[schema_id] = schema
+
+        if subject:
+            self._add_to_cache(self.subject_to_schema_ids,
+                               subject, schema, schema_id)
+            if version:
+                self._add_to_cache(self.subject_to_schema_versions,
+                                   subject, schema, version)
+
+    def register(self, subject, avro_schema):
+        """
+        Register a schema with the registry under the given subject
+        and receive a schema id.
+
+        avro_schema must be a parsed schema from the python avro library
+
+        Multiple instances of the same schema will result in cache misses.
+
+        @:param: subject: subject name
+        @:param: avro_schema: Avro schema to be registered
+        @:returns: schema_id: int value
+        """
+
+        schemas_to_id = self.subject_to_schema_ids.get(subject, {})
+
+        schema_id = schemas_to_id.get(avro_schema, None)
+        if schema_id != None:
+            return schema_id
+        # send it up
+        url = '/'.join([self.url, 'subjects', subject, 'versions'])
+        # body is { schema : json_string }
+
+        body = {'schema': json.dumps(avro_schema.to_json())}
+        result, code = self._send_request(url, method='POST', body=body)
+        if (code != 200):
+            self.log.error("Unable to register schema. Error code:" + str(code))
+            raise ClientError("Unable to register schema. Error code:" + str(code))
+        # result is a dict
+        schema_id = result['id']
+        # cache it
+        self._cache_schema(avro_schema, schema_id, subject)
+        return schema_id
+
+    def get_by_id(self, schema_id):
+        """Retrieve a parsed avro schema by id or None if not found
+        @:param: schema_id: int value
+        @:returns: Avro schema
+        """
+        if schema_id in self.id_to_schema:
+            return self.id_to_schema[schema_id]
+        # fetch from the registry
+        url = '/'.join([self.url, 'schemas', 'ids', str(schema_id)])
+
+        result, code = self._send_request(url)
+        if (code != 200):
```

Comment:
nit: May want to explicitly check for 404 vs other status codes and log an error where appropriate.